src/System.Private.CoreLib/shared/System/Threading/Mutex.cs

```diff
@@ -0,0 +1,218 @@
+            }
+#endif
+        }
+
+        private void CreateMutexCore(bool initiallyOwned, string name, out bool createdNew)
+        {
+#if !PLATFORM_UNIX
+            Debug.Assert(name == null || name.Length <= Interop.Kernel32.MAX_PATH);
+#endif
+
+#if CORECLR
+            uint mutexFlags = initiallyOwned ? Win32Native.CREATE_MUTEX_INITIAL_OWNER : 0;
+            SafeWaitHandle mutexHandle = Win32Native.CreateMutexEx(null, name, mutexFlags, AccessRights);
+#else
+            uint mutexFlags = initiallyOwned ? (uint)Interop.Constants.CreateMutexInitialOwner : 0;
+            SafeWaitHandle mutexHandle = Interop.mincore.CreateMutexEx(IntPtr.Zero, name, mutexFlags, AccessRights);
+#endif
+            int errorCode = Marshal.GetLastWin32Error();
+
+            if (mutexHandle.IsInvalid)
+            {
+                mutexHandle.SetHandleAsInvalid();
+#if PLATFORM_UNIX
+                if (errorCode == Interop.Errors.ERROR_FILENAME_EXCED_RANGE)
+                    // On Unix, length validation is done by CoreCLR's PAL after converting to utf-8
+                    throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name, WaitHandleNameMax), nameof(name));
+#endif
+                if (errorCode == Interop.Errors.ERROR_INVALID_HANDLE)
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
+#if CORECLR
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+#else
+                throw ExceptionFromCreationError(errorCode, name);
+#endif
+            }
+
+            createdNew = errorCode != Interop.Errors.ERROR_ALREADY_EXISTS;
+            SafeWaitHandle = mutexHandle;
+        }
+
+        private static OpenExistingResult OpenExistingWorker(string name, out Mutex result)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (name.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_EmptyName, nameof(name));
+            }
+
+            VerifyNameForCreate(name);
+            result = null;
+            // To allow users to view & edit the ACL's, call OpenMutex
+            // with parameters to allow us to view & edit the ACL.  This will
+            // fail if we don't have permission to view or edit the ACL's.  
+            // If that happens, ask for less permissions.
+#if CORECLR            
+            SafeWaitHandle myHandle = Win32Native.OpenMutex(AccessRights, false, name);
+#else
+            SafeWaitHandle myHandle = Interop.mincore.OpenMutex(AccessRights, false, name);
+#endif
+            if (myHandle.IsInvalid)
+            {
+                int errorCode = Marshal.GetLastWin32Error();
+
+#if PLATFORM_UNIX
+                if (name != null && errorCode == Interop.Errors.ERROR_FILENAME_EXCED_RANGE)
```

Comment:
It may be nice to throw to move Windows on the same plan as Unix here - throw the name length error only after the OS returns the name is too long. It is the plan we are trying to be on everywhere else: don't second guess the name length limits in the OS.